File: search_rotated_sorted.py
# https://leetcode.com/problems/search-in-rotated-sorted-array/

class Solution: # 40ms, linear time
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        if len(nums) == 0:
            return -1        
        elif target < nums[0]:
            i = len(nums) - 1
            inc = -1
        elif target > nums[0]:
            i = 0
            inc = 1
        else:
            return 0
        
        while 0 <= i <= len(nums)-1:
            if target == nums[i]:
                return i
            i += inc
        
        return -1


# Wayy easier solution could have been binary search: 36ms O(log(n))

# class Solution:
#     def search(self, nums, target):
#         """
#         :type nums: List[int]
#         :type target: int
#         :rtype: int
#         """
#         def find_rotate_index(left, right):
#             if nums[left] < nums[right]:
#                 return 0
            
#             while left <= right:
#                 pivot = (left + right) // 2
#                 if nums[pivot] > nums[pivot + 1]:
#                     return pivot + 1
#                 else:
#                     if nums[pivot] < nums[left]:
#                         right = pivot - 1
#                     else:
#                         left = pivot + 1
                
#         def search(left, right):
#             """
#             Binary search
#             """
#             while left <= right:
#                 pivot = (left + right) // 2
#                 if nums[pivot] == target:
#                     return pivot
#                 else:
#                     if target < nums[pivot]:
#                         right = pivot - 1
#                     else:
#                         left = pivot + 1
#             return -1
        
#         n = len(nums)
        
#         if n == 0:
#             return -1
#         if n == 1:
#             return 0 if nums[0] == target else -1 
        
#         rotate_index = find_rotate_index(0, n - 1)
        
#         # if target is the smallest element
#         if nums[rotate_index] == target:
#             return rotate_index
#         # if array is not rotated, search in the entire array
#         if rotate_index == 0:
#             return search(0, n - 1)
#         if target < nums[0]:
#             # search on the right side
#             return search(rotate_index, n - 1)
#         # search on the left side
#         return search(0, rotate_index)

File: search_rotated_sorted.cpp
class Solution { // 4ms, faster than 99.40% of cpp submissions. Based on quicksort / selection idea by leetcode.
public:
    int search(vector<int>& nums, int target) {
        // look for rotation index
        if (nums.size() == 0){
            return -1;
        } 
        else if (nums.size() == 1){
            return (nums[0] == target ? 0 : -1);
        } 
        else if (nums.size() == 2){
            return (nums[0] == target ? 0 : (nums[1] == target ? 1 : -1));
        }
        
        int left = 0;
        int right = nums.size() - 1;
        int mid = (left+right) / 2;
        if(nums[left] < nums[right]){
            // already sorted
            mid = 0;
        }
        else {
            while(left <= right){
                // cout<<left<<" "<<right<<endl;
                mid = (left + right) / 2;
                if (nums[mid] > nums[mid+1]){
                    break;
                }
                else if (nums[mid] >= nums[left]){
                    left = mid+1;
                } 
                else {
                    right = mid-1;
                }
            }
            mid += 1;
        }
        // cout << mid << endl;
        // assuming left == right never happens
        if (target >= nums[mid] and target <= nums[nums.size()-1]){
            left = mid;
            right = nums.size() - 1;
        } else {
            left = 0;
            right = mid-1;
        }
        // cout<<left<<" "<<right<<endl;
        mid = (left + right) / 2;
        while(left <= right){
            // cout<<left<<" "<<right<<endl;
            mid = (left + right) / 2;
            if (nums[mid] == target){
                break;
            }
            if (target < nums[mid]){
                right = mid-1;
            } 
            else {
                left = mid+1;
            }
        }
        
        return (nums[mid] == target ? mid : -1);
        
    }
};